### PR TITLE
Add tensorflow tensor support

### PR DESCRIPTION
**Pull request by darshil929**
### Changes
- Added support for `ord=0` in tf_linalg.py's norm function 
- Fixed the keepdims parameter for nuclear norm case
- Added tests that check norm with keepdims=True for both CPU and GPU classes
- Added tests for lstsq function with multiple right-hand sides (rank-2 tensors)

### Reason for changes
Right now TensorFlow tensors don't work with nncf.Tensor, which is blocking issue #[3041](https://github.com/openvinotoolkit/nncf/pull/3041) from being done. This PR adds TensorFlow backend support so we can use TensorFlow tensors with NNCF operations. I'm continuing the work from PR #[3106](https://github.com/openvinotoolkit/nncf/pull/3106) but fixing some things reviewers pointed out and adding stuff that was missing.

### Related tickets
#[3041](https://github.com/openvinotoolkit/nncf/pull/3041) 
#[3106](https://github.com/openvinotoolkit/nncf/pull/3106) 

### Tests
- Added a test_norm_keepdims function to both CPU and GPU classes to check if keepdims works right
- Added test_lstsq_rank2 to make sure least squares works with multiple columns
- Note: I've added the requested tests, but encountered environment setup issues that prevented me from running them successfully. The error appears to be related to function signature registration in the TensorFlow backend integration, which I think is beyond the scope of my changes

### CODE REVIEW

**Comment by darshil929** [comment]
@kshpv 

I've implemented the suggested change to the save_file.register decorator. I've also fixed string literals in exceptions in tf_linalg.py and tf_numeric.py to address the pre-commit linting issues that were causing CI failures. The PR should be ready for review now.

File: nncf/tensor/tensor.py [hunk 1]
```diff
@@ -140,7 +140,7 @@ def __floordiv__(self, other: Union[Tensor, T_NUMBER]) -> Tensor:
     def __rfloordiv__(self, other: Union[Tensor, T_NUMBER]) -> Tensor:
         return cast(Tensor, _call_function("_binary_reverse_op_nowarn", self, other, operator.floordiv))
 
-    def __ifloordiv__(self, other: Union[Tensor, T_NUMBER]) -> Tensor:
```
Comment (kshpv) [inline nncf/tensor/tensor.py:143]: Please, rollback these changes

File: nncf/version.py [hunk 2]
```diff
@@ -9,7 +9,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-__version__ = "2.16.0"
```
Comment (kshpv) [inline nncf/version.py:12]: Please, rollback these changes

File: nncf/tensor/functions/tf_io.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+from nncf.tensor import TensorDeviceType
+from nncf.tensor.functions import io as io
+
+
+def load_file(file_path: str, *, device: Optional[TensorDeviceType] = None) -> Dict[str, tf.Tensor]:
+    return tf_load_file(file_path)
+
+
+@io.save_file.register
+def _(data: Dict[str, tf.Tensor], file_path: str) -> None:
```
Comment (kshpv) [inline nncf/tensor/functions/tf_io.py:27]:
```suggestion
def _(data: Dict[str, tf.Tensor], file_path: Path) -> None:
```

File: nncf/tensor/functions/tf_io.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+import tensorflow as tf
+from safetensors.tensorflow import load_file as tf_load_file
+from safetensors.tensorflow import save_file as tf_save_file
+
+from nncf.tensor import TensorDeviceType
+from nncf.tensor.functions import io as io
+
+
+def load_file(file_path: str, *, device: Optional[TensorDeviceType] = None) -> Dict[str, tf.Tensor]:
```
Comment (kshpv) [inline nncf/tensor/functions/tf_io.py:22]:
```suggestion
def load_file(file_path: Path, *, device: Optional[TensorDeviceType] = None) -> Dict[str, tf.Tensor]:
```

File: tests/cross_fw/test_templates/template_test_nncf_tensor.py [hunk 5]
```diff
@@ -113,7 +113,8 @@ def test_operators_tensor(self, op_name):
         assert res.dtype == res_nncf.data.dtype
         assert all(res == res_nncf.data)
         assert isinstance(res_nncf, Tensor)
-        assert res_nncf.device == nncf_tensor_a.device
```
Comment (kshpv) [inline tests/cross_fw/test_templates/template_test_nncf_tensor.py:116]: Can we avoid these changes by using in an function device conversion via? 
I think for some cases this can help to return a tensor on a specified device
```
with tf.device(device):
  tensor = tf.identity(tensor)
```

Comment (darshil929) [reply tests/cross_fw/test_templates/template_test_nncf_tensor.py:116]: I've implemented the device conversion pattern you recommended by adding tf.identity() inside the device context for tensor operations in tf_numeric.py to ensure tensors remain on the specified device.

removed conditional check in the test.

**Comment by darshil929** [comment]
> > @kshpv
> > I've implemented the suggested change to the save_file.register decorator. I've also fixed string literals in exceptions in tf_linalg.py and tf_numeric.py to address the pre-commit linting issues that were causing CI failures. The PR should be ready for review now.
> 
> Hi @darshil929 Could you please fix precommit issues at first?

I'm analyzing the precommit issues now and will get them fixed as soon as possible.

File: nncf/tensor/functions/tf_linalg.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import warnings
+from typing import Optional, Tuple, Union
+
+import tensorflow as tf
```
Comment (darshil929) [inline nncf/tensor/functions/tf_linalg.py:15]: @kshpv 

Sir, I notice mypy is showing errors about missing TensorFlow type stubs (Cannot find implementation or library stub for module named "tensorflow"). Would adding # type: ignore to the TensorFlow imports be the appropriate fix, or is there a more preferred way to handle these in the NNCF codebase?

Comment (kshpv) [reply nncf/tensor/functions/tf_linalg.py:15]: To fix this issue it's needed to include `tensorflow` package for mypy precommit environment  - https://github.com/openvinotoolkit/nncf/blob/develop/.github/workflows/mypy.yml#L26
Just add this line 
`pip install -e . tensorflow -c constraints.txt`
after this:
`pip install -e . torch -c constraints.txt`

File: nncf/tensor/functions/tf_linalg.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+from nncf.tensor.functions import linalg
+
+
+@linalg.norm.register(tf.Tensor)
```
Comment (darshil929) [inline nncf/tensor/functions/tf_linalg.py:20]: @kshpv 

Sir, I notice there are test failures related to function arguments in the norm function:

`ERROR tests/tensorflow/test_weekly.py - RuntimeError: Differ names of arguments for dispatched a function norm in tensorflow.python.framework.tensor.
Expected: ['a', 'ord', 'axis', 'keepdims'] []
Actual:   [] []`

Would removing `tf.Tensor` from the decorator (changing from `@linalg.norm.register(tf.Tensor)` to just `@linalg.norm.register`) solve this issue, similar to the fix you suggested for the save_file function?

Comment (kshpv) [reply nncf/tensor/functions/tf_linalg.py:20]: yes

Comment (darshil929) [reply nncf/tensor/functions/tf_linalg.py:20]: And should I apply the same change to all function decorators in the tf_linalg.py file?

Comment (kshpv) [reply nncf/tensor/functions/tf_linalg.py:20]: I think so. Your reference is here https://github.com/openvinotoolkit/nncf/blob/develop/nncf/tensor/functions/numpy_linalg.py

**Comment by darshil929** [comment]
Hi @kshpv 

Sir, I've made several changes to fix the type annotations and error handling issues in the backend implementation that were causing the mypy and pytest/tensorflow CI checks to fail. The changes include:

- Fixed missing return paths in device detection function
- Added proper type checking for axis parameters in various functions
- Fixed return type annotation mismatches
- Added proper type parameters to Callable and ndarray generic types
- Fixed device mapping type compatibility issues in tensor creation functions
- Added explicit type handling for expand_dims function to prevent iteration errors

Could you please review these changes and rerun the CI checks to verify they resolve the issues? Please let me know if any additional adjustments are needed.

Thank you!

**Comment by kshpv** [comment]
> Hi @kshpv
> 
> Sir, I've made several changes to fix the type annotations and error handling issues in the backend implementation that were causing the mypy and pytest/tensorflow CI checks to fail. The changes include:
> 
> * Fixed missing return paths in device detection function
> * Added proper type checking for axis parameters in various functions
> * Fixed return type annotation mismatches
> * Added proper type parameters to Callable and ndarray generic types
> * Fixed device mapping type compatibility issues in tensor creation functions
> * Added explicit type handling for expand_dims function to prevent iteration errors
> 
> Could you please review these changes and rerun the CI checks to verify they resolve the issues? Please let me know if any additional adjustments are needed.
> 
> Thank you!

I recommend you run tests locally, it speeds up your development. Please, take a look at failed precommit builds, you need to focus on fixing them and ensure that failed tests pass locally. After that I can make a full-fledged review

**Comment by darshil929** [comment]
> > Hi @kshpv
> > Sir, I've made several changes to fix the type annotations and error handling issues in the backend implementation that were causing the mypy and pytest/tensorflow CI checks to fail. The changes include:
> > 
> > * Fixed missing return paths in device detection function
> > * Added proper type checking for axis parameters in various functions
> > * Fixed return type annotation mismatches
> > * Added proper type parameters to Callable and ndarray generic types
> > * Fixed device mapping type compatibility issues in tensor creation functions
> > * Added explicit type handling for expand_dims function to prevent iteration errors
> > 
> > Could you please review these changes and rerun the CI checks to verify they resolve the issues? Please let me know if any additional adjustments are needed.
> > Thank you!
> 
> I recommend you run tests locally, it speeds up your development. Please, take a look at failed precommit builds, you need to focus on fixing them and ensure that failed tests pass locally. After that I can make a full-fledged review

Yes sir, I understand. I've been experiencing some environment issues with local test execution, but I'll definitely get those resolved first. I'll update you as soon as I have tested everything locally and verified the fixes.

**Comment by darshil929** [comment]
Hey! @kshpv 

I've fixed up a bunch of issues in the TensorFlow backend code that were making our mypy checks and TensorFlow tests fail in CI. I added proper implementations for basic operations like add/subtract/multiply, fixed the norm function to handle default values better, and made tensor loading work correctly with device placement.

Also fixed some security stuff like preventing tensor saves to symlinks, and cleaned up a lot of type annotations that were causing mypy to complain. The keepdims parameter was causing weird results in some cases, so I fixed that too.

I've run everything locally and all tests pass now! The TensorFlow backend should work properly with all the NNCF operations now.

Could you take a look when you get a chance? Let me know if you spot anything else that needs fixing.

Thanks!

**Comment by darshil929** [comment]
Hey @kshpv 

While working on fixing the failing CI checks, I discovered that the ONNX quantization tests are failing due to dataset download issues:

pytest tests/onnx/quantization/test_ptq_regression.py::test_compression[mobilenetv2-12]
pytest tests/onnx/quantization/test_ptq_regression.py::test_compression[resnet50-v1-7]
pytest tests/onnx/quantization/test_ptq_regression.py::test_compression[efficientnet-lite4-11]

These failures are unrelated to my TensorFlow tensor implementation work and occur when downloading/extracting the dataset files. Should I address these issues too? or focus only on the TensorFlow-related changes?

**Comment by darshil929** [comment]
Hey! @kshpv 

I've fixed the issue that was causing the CI checks to fail. The problem was simply a missing backend function implementation for float type in the TensorFlow code. I've added the missing implementation, and now all the checks are passing successfully.

I apologize for the inconvenience. Could you please review the changes when you get a chance?

Thank you!

File: nncf/tensor/tensor.py [hunk 8]
```diff
@@ -85,6 +85,10 @@ def __len__(self) -> int:
     # built-in operations
 
     def __add__(self, other: Union[Tensor, T_NUMBER]) -> Tensor:
+        if self.backend == TensorBackend.tf:
```
Comment (kshpv) [inline nncf/tensor/tensor.py:88]: Couldn't the logic for the `TensorFlow` backend be encapsulated in the same way as it is for `PyTorch` and `NumPy`? What's the reasoning behind these changes?

Comment (darshil929) [reply nncf/tensor/tensor.py:88]: These changes were implemented specifically to address the test failure where TensorFlow operations weren't preserving device information, leading to assertions failing with:

`assert res_nncf.device == nncf_tensor_a.device
AssertionError`

The test expected the result tensor to maintain the same device as the input tensor after operations, but this wasn't happening with TensorFlow's default behavior. To explicitly enforce device preservation within each operation by using TensorFlow's device context management, I routed TensorFlow tensor operations through the numeric dispatcher system.

Comment (kshpv) [reply nncf/tensor/tensor.py:88]: You should not fix the issue with `tensorflow` backend in a common code. Instead it should be done inside of the `tensorflow` functions using this approach where possible - https://github.com/openvinotoolkit/nncf/pull/3340#discussion_r1998300114 If it does not fix the device issue I think we can do a workaround in tests like this - https://github.com/openvinotoolkit/nncf/pull/3340/files#diff-34b3b29becfe5a7b6e31fa3c7e73e802f1ee0d4853286bb28307c2890ef5b048R132 but I also want to hear the opinion from @alexsu52 on this

Comment (darshil929) [reply nncf/tensor/tensor.py:88]: I understand your concern about fixing the issue by making the changes in common code and I surely won't repeat it in the future changes.

I did initially try the approach from [comment](https://github.com/openvinotoolkit/nncf/pull/3340#discussion_r1998300114), implementing device context within TensorFlow functions, but tests  still continued failing with the same device assertion error.

Comment (kshpv) [reply nncf/tensor/tensor.py:88]: To address device-related issues in `TensorFlow`, you should not only use the `with tf.device(device)`: context to specify the device for operations, but also apply `tf.identity(result)` to the resulting tensor before returning it. This approach can help ensure that the operations are executed on the intended device. 
For example for operation `tf.eye(...)` the following code place the `a` tensor to 'GPU' while without `tf.identity` `a` will be on 'CPU':
```
with tf.device('GPU'):
    a = tf.eye(2, dtype=tf.int32)
    a = tf.identity(a)
```

Comment (darshil929) [reply nncf/tensor/tensor.py:88]: Okay got it!

`@numeric.add.register
def _(x1: tf.Tensor, x2: Union[tf.Tensor, float]) -> tf.Tensor:
    with tf.device(x1.device):
        if not isinstance(x2, tf.Tensor):
            x2 = tf.convert_to_tensor(x2, dtype=x1.dtype)
        result = tf.add(x1, x2)
        return tf.identity(result)`

Is this implementation of add function correct based on your example?

Comment (kshpv) [reply nncf/tensor/tensor.py:88]: In the method `def __add__(self, other: Union[Tensor, T_NUMBER])`, the addition operation is performed on `self.data` and `unwrap_tensor_data(other)`. Since these are `TensorFlow` tensors, the `+` operator invokes the original `TensorFlow`addition operation, which does not specify a device. Consequently, the result may not be returned on the same device as the original tensors.

Therefore, in the test cases for `Tensor` operators, there should be special handling for the `TensorFlow` backend. Additionally, it's important to include a comment explaining that the original `TensorFlow` operation does not guarantee the result will be on the same device as the input tensors.

Please update the tests for the `TensorFlow` backend to skip device checks, and add a comment explaining why the device check is skipped due to `TensorFlow's` default behavior.

Thanks! 
cc @alexsu52

Comment (darshil929) [reply nncf/tensor/tensor.py]: Thanks for the detailed explanation! I've updated the test cases to skip device consistency checks for the `TensorFlow` backend and added a comment explaining TensorFlow's behavior regarding tensor arithmetic operations. I have also reverted back the functions in `TensorFlow` backend as they were originally. 

Please do let me know if any further refinements are needed!

File: nncf/tensor/functions/tf_numeric.py [hunk 9]
```diff
@@ -0,0 +1,619 @@
...
+    m: Optional[int] = None,
+    *,
+    dtype: Optional[TensorDataType] = None,
+    device: Optional[TensorDeviceType] = None,
+) -> tf.Tensor:
+    tf_dtype = DTYPE_MAP[dtype] if dtype is not None else None
+    tf_device = DEVICE_MAP[device] if device is not None else None
+    p_args = (n,) if m is None else (n, m)
+    with tf.device(tf_device):
+        return tf.eye(*p_args, dtype=tf_dtype)
```
Comment (kshpv) [inline nncf/tensor/functions/tf_numeric.py:571]: This should resolve `test_fn_eye()` device problem
```suggestion
        eye = tf.eye(*p_args, dtype=tf_dtype)
        return tf.identity(eye)
```

**Review by kshpv — COMMENTED** [review]
Thank you! Please, take a look at my comments. Some functions look bulky but they pass the tests, that's okay for me

File: nncf/tensor/functions/numeric.py [hunk 10]
```diff
@@ -105,6 +105,16 @@ def abs(a: Tensor) -> Tensor:
     """
 
 
+@tensor_dispatcher
+def neg(a: Tensor) -> Tensor:
```
[the remark this answers is not in the document]
Comment (darshil929) [reply nncf/tensor/functions/numeric.py:109]: I was facing this error:
```
AttributeError: module 'nncf.tensor.functions.numeric' has no attribute 'neg'
```
that led me to adding these functions.
But after solving device preservation check issue, i think this error is resolved too so,  i have rolled back these changes.

File: tests/cross_fw/test_templates/template_test_nncf_tensor.py [hunk 11]
```diff
@@ -1550,6 +1621,8 @@ def test_fn_eye(self, n, m, ref):
                 ]
             ):
                 continue
+            if (not dtype.is_float()) and self.backend() == TensorBackend.tf and self.device() == TensorDeviceType.GPU:
+                continue
             tensor_a = fns.eye(n, m, backend=self.backend(), dtype=dtype, device=self.device())
             assert isinstance(tensor_a, Tensor)
             assert tensor_a.device == self.device()
```
[the remark this answers is not in the document]
Comment (darshil929) [reply tests/cross_fw/test_templates/template_test_nncf_tensor.py:1628]: In this test,
`assert tensor_a.device == self.device()`
has successfully passed everytime. Should I still update this assertion?

If Yes, should I make the similar updates to `test_fn_zeros` and `test_fn_arange` functions too?

File: tests/cross_fw/test_templates/template_test_nncf_tensor.py [hunk 12]
```diff
@@ -113,7 +113,15 @@ def test_operators_tensor(self, op_name):
...
-        assert res_nncf.device == nncf_tensor_a.device
+        # TensorFlow's default tensor operations (e.g., addition, subtraction, multiplication, power, negation, etc.)
+        # do not guarantee that the resulting tensor will be on the same device as the input tensors. This is because
+        # TensorFlow does not explicitly specify a device when performing these operations. As a result, the output
+        # tensor may be placed on a different device than expected.
+        #
+        # To handle this behavior, we skip device consistency checks for the TensorFlow backend in test cases
+        # involving tensor operations.
+        if self.backend() != TensorBackend.tf:
+            assert res_nncf.device == nncf_tensor_a.device
```
[the remark this answers is not in the document]
Comment (darshil929) [reply tests/cross_fw/test_templates/template_test_nncf_tensor.py:124]: `test_operators_tensor` is the only function where this check is failing currently and I have updated it according to your suggestion.

**Comment by darshil929** [comment]
Hi @kshpv 

I've addressed all your comments in my latest commit.

I had a question about the [comment](https://github.com/openvinotoolkit/nncf/pull/3340#discussion_r2014170535), could you please clarify it?

Also, what should be the next steps for this PR? Are there any other improvements or changes you'd like me to make further?

Thank you for your detailed feedback and guidance!

File: tests/cross_fw/test_templates/template_test_nncf_tensor.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (kshpv) [reply tests/cross_fw/test_templates/template_test_nncf_tensor.py:1623]: You should remove at first
```
            if (not dtype.is_float()) and self.backend() == TensorBackend.tf and self.device() == TensorDeviceType.GPU:
                continue
```

File: nncf/tensor/functions/numeric.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/tensor/functions/numeric.py]: Please remove all added funcitons

File: tests/cross_fw/test_templates/template_test_nncf_tensor.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (kshpv) [reply tests/cross_fw/test_templates/template_test_nncf_tensor.py]: This PR changes several tests with the check on Tensorflow backend. Please align all of them with the suggestion above

**Review by kshpv — COMMENTED** [review]
Thank you for your hard work! I have no more questions regarding these changes.
Waiting for review from @alexsu52

**Comment by kshpv** [comment]
In the meantime you can continue working on finalizing the https://github.com/openvinotoolkit/nncf/issues/3041 using your changes

**Comment by darshil929** [comment]
Thank you! 

Sir, I have a few minor variable naming changes I missed in my latest commit. Should I add these after the current pytest run is complete?

**Comment by darshil929** [comment]
> In the meantime you can continue working on finalizing the #3041 using your changes

Sure! i will start working on it immediately.

**Comment by kshpv** [comment]
> > Thank you for your hard work! I have no more questions regarding these changes. Let's wait for a review from @alexsu52
> 
> Thank you!
> 
> Sir, I have a few minor variable naming changes I missed in my latest commit. Should I add these after the current pytest run is complete?

Yes, please continue working on this. I will take another look if necessary

**Comment by darshil929** [comment]
@kshpv 

Sir, I've made some minor variable name changes and re-applied `tf.identity()` to the functions that were facing device preservation issues initially to maintain consistency in code. Sorry for the delay.

**Comment by darshil929** [comment]
Am I supposed to create a new PR to complete the work on issue [#3041](https://github.com/openvinotoolkit/nncf/issues/3041) or it should be done in the current PR?

**Comment by kshpv** [comment]
> Am I supposed to create a new PR to complete the work on issue [#3041](https://github.com/openvinotoolkit/nncf/issues/3041) or it should be done in the current PR?

It would be easier for reviewers to split the work into parts. I suggest creating a new PR

**Comment by darshil929** [comment]
> > Am I supposed to create a new PR to complete the work on issue [#3041](https://github.com/openvinotoolkit/nncf/issues/3041) or it should be done in the current PR?
> 
> It would be easier for reviewers to split the work into parts. I suggest creating a new PR

Hi @kshpv 

I'm currently working on preparing my proposal draft for GSOC project 14 - "Accelerating Inference of NNCF-Compressed LLMs with Triton," for OpenVINO-NNCF. I'll raise the PR for issue #3041 as soon as my proposal work is completed, which should be within a couple of days.
I would really appreciate a review of my proposal and suggestions for improvement. What would be the best way to share my Google Doc with you? Would you prefer I send it to your email, or should I contact you through Discord instead?

Thank you for your time and guidance.

**Comment by alexsu52** [comment]
> > > Am I supposed to create a new PR to complete the work on issue [#3041](https://github.com/openvinotoolkit/nncf/issues/3041) or it should be done in the current PR?
> > 
> > 
> > It would be easier for reviewers to split the work into parts. I suggest creating a new PR
> 
> Hi @kshpv
> 
> I'm currently working on preparing my proposal draft for GSOC project 14 - "Accelerating Inference of NNCF-Compressed LLMs with Triton," for OpenVINO-NNCF. I'll raise the PR for issue #3041 as soon as my proposal work is completed, which should be within a couple of days. I would really appreciate a review of my proposal and suggestions for improvement. What would be the best way to share my Google Doc with you? Would you prefer I send it to your email, or should I contact you through Discord instead?
> 
> Thank you for your time and guidance.

You can send your proposal to [alexander.suslov@intel.com](mailto:alexander.suslov@intel.com)

**Comment by darshil929** [comment]
> > > > Am I supposed to create a new PR to complete the work on issue [#3041](https://github.com/openvinotoolkit/nncf/issues/3041) or it should be done in the current PR?
> > > 
> > > 
> > > It would be easier for reviewers to split the work into parts. I suggest creating a new PR
> > 
> > 
> > Hi @kshpv
> > I'm currently working on preparing my proposal draft for GSOC project 14 - "Accelerating Inference of NNCF-Compressed LLMs with Triton," for OpenVINO-NNCF. I'll raise the PR for issue #3041 as soon as my proposal work is completed, which should be within a couple of days. I would really appreciate a review of my proposal and suggestions for improvement. What would be the best way to share my Google Doc with you? Would you prefer I send it to your email, or should I contact you through Discord instead?
> > Thank you for your time and guidance.
> 
> You can send your proposal to [alexander.suslov@intel.com](mailto:alexander.suslov@intel.com)

Sure sir, I am working on a example-prototype to add as a part of my proposal and demonstrate my understandings which will take a little time to complete. As soon as it is done, I will email you the draft.

**Comment by darshil929** [comment]
> > @kshpv
> > Sir, I've made some minor variable name changes and re-applied `tf.identity()` to the functions that were facing device preservation issues initially to maintain consistency in code. Sorry for the delay.
> 
> Could you check that adding `tf.identity()` resolves device issues? For all tests where it resolves, please remove the skipping device check. If it does not resolve, then please roll back adding `tf.identity()`

Sir, I have made the required changes.

**Review by kshpv — APPROVED** [review]

**Comment by darshil929** [comment]
Hi @kshpv  @alexsu52 

Sir, after investigating the tensor statistics code, here's my understanding of the situation and my planned approach:

What I found:
- As mentioned in issue description, we have two implementations: old one in `common/tensor_statistics` and new one in `experimental/common/tensor_statistics`
- There's already some circular dependencies between them - the old code is importing `TensorStatistic` and `TensorCollector` from experimental, while experimental is importing `ReductionAxes` from the old code
- New implementation has more test coverage (25 test files compared to 11 for old implementation)
- New implementation seems to be more modular with separate reducers and aggregators instead of monolithic collectors

My planned approach:
1. First break the circular dependency by replacing the `ReductionAxes` import in experimental with its own `AggregationAxes`
2. Start moving functionality in this order:
   - Statistics classes first
   - Then collectors (mapping old collectors to appropriate reducer+aggregator combinations)
   - Then serialization and validation
3. Update framework-specific implementations (PyTorch, TensorFlow, ONNX, OpenVINO)
4. Update algorithms that use tensor statistics
5. Final step: replace contents in common directory with experimental implementation and remove the experimental directory

Questions:
- Is my understanding correct?
- Are there any specific components I should be careful with during migration?
- Do you agree with my step-by-step approach or would you suggest a different order?
- Any pitfalls I should watch out for based on previous migration attempts?

I'm planning to split this into multiple PRs to make reviews easier. Please let me know if this approach makes sense or if you have alternative suggestions!

File: nncf/tensor/functions/tf_linalg.py [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+from typing import Literal, Optional, Tuple, Union
+
+import tensorflow as tf
+
+from nncf.tensor.functions import linalg
+
+
+@linalg.norm.register
+def _(
+    a: tf.Tensor,
```
Comment (alexsu52) [inline nncf/tensor/functions/tf_linalg.py:22]: I would recommenced to use test driven approach for fixing functions implementation.  I found several bugs in the implementation, one of them:
{50B0C437-8736-4A51-BD88-8181BB1D362F}

I suggest the following steps to fix the implementation:
- expand the test for this function by adding missing parameter combinations to the test
- based on the test, correct the implementation of the function

**Review by alexsu52 — CHANGES_REQUESTED** [review]

File: docs/api/source/conf.py [hunk 14]
```diff
@@ -145,8 +145,9 @@ def collect_api_entities() -> APIInfo:
     "nncf.tensor.functions.torch_linalg",
     "nncf.tensor.functions.torch_io",
     "nncf.tensor.functions.numpy_io",
-    "nncf.tensor.functions.openvino_numeric",
-    "nncf.torch.dynamic_graph.patch_pytorch",
```
[the remark this answers is not in the document]
Comment (darshil929) [reply docs/api/source/conf.py:149]: These modules were likely added to the develop branch while the PR #3106 was sitting idle. When I rebased against the PR branch, these newer additions got dropped. I'll restore them. It was my mistake that I overlooked these changes.

**Comment by darshil929** [comment]
Hi @alexsu52 
Sir, I've implemented comprehensive test coverage for the norm function as requested:

- Added tests for all norm types on tensors from 1D to 4D
- Fixed the bug with nuclear norm and keepdims parameter for higher dimensions
- Added tests for ord=0 implementation
- Included edge case tests for empty tensors and extreme values (NaN, Inf)

All tests now pass. Could you please review when you have a chance.
Thank you!

File: tests/tensorflow/test_tensor.py [hunk 15]
```diff
@@ -87,6 +87,371 @@ def test_lstsq_rank2(self):
                 expected_val = expected[i][j]
                 assert abs(x_val - expected_val) < 0.2, f"Value at ({i},{j}) is {x_val}, expected {expected_val}"
 
+    def test_norm_comprehensive(self):
```
[the remark this answers is not in the document]
Comment (darshil929) [reply tests/tensorflow/test_tensor.py:90]: Yes, I can move these norm tests to the template class in template_test_nncf_tensor.py. This will make them run on all tensor backends, ensuring consistent norm functionality across frameworks.

I will commit these changes as soon as I can.

**Comment by darshil929** [comment]
Hi @alexsu52 

Sir, I've successfully moved norm tests to the TemplateTestNNCFTensorOperators class. All tests are running perfectly on the TensorFlow backend.

However, there are some failures on PyTorch and NumPy backends.

PyTorch backend failures:
Pasted image 20250504022947

NumPy backend failures:
Pasted image 20250504023028

Could you please advice if I should modify the backend implementations to handle these edge cases?
Or is there another approach you'd recommend?

Thank you!

**Comment by alexsu52** [comment]
> Hi @alexsu52
> 
> Sir, I've successfully moved norm tests to the TemplateTestNNCFTensorOperators class. All tests are running perfectly on the TensorFlow backend.
> 
> However, there are some failures on PyTorch and NumPy backends.
> 
> PyTorch backend failures: Pasted image 20250504022947
> 
> NumPy backend failures: Pasted image 20250504023028
> 
> Could you please advice if I should modify the backend implementations to handle these edge cases? Or is there another approach you'd recommend?
> 
> Thank you!

Hi @darshil929,

Thank you for the contribution!

Please, update tests regarding the description of the norm function https://github.com/alexsu52/nncf/blob/develop/nncf/tensor/functions/linalg.py#L19. For example, the test `test_norm_order_zero` failed calling `result = fns.linalg.norm(tensor_2d, ord=0)`, but the norm function does not supports ord=0 for 2d tensor https://github.com/alexsu52/nncf/blob/develop/nncf/tensor/functions/linalg.py#L38.

Some insights, numpy behavior is the default behavior unless otherwise noted. If you find a bug in the numpy, pytorch or openvino backends, please feel free to fix it. Please also keep the function description up to date.

**Review by alexsu52 — COMMENTED** [review]
Please, fix the pre-commit tests.

**Comment by darshil929** [comment]
Sir,

I've updated the norm tests to align with the documented behavior in `linalg.py` where `ord=0` is officially supported only for vectors, not matrices. The tests now properly validate vector norm behavior while maintaining compatibility across all backends (TensorFlow, PyTorch, and NumPy).

I've also preserved the TensorFlow implementation's extended capabilities for `ord=0`

All tests are now passing. Could you please review these changes?

Thanks!

**Review by alexsu52 — APPROVED** [review]
LGTM
